### PR TITLE
Adding Symfony Empty Edition

### PR DESCRIPTION
**Pull request by gnugat**
The [Symfony Empty Edition](https://github.com/gnugat/symfony-empty-edition) is a blank canvas: it includes the strict minimum (`FrameworkBundle`) to allow you to build a custom made application with only the dependencies you need.

### CODE REVIEW

**Comment by mheki** [comment]
Good stuff :+1:  For many simple projects you don't really need standard edition to start with.

**Comment by OskarStark** [comment]
:-1:

**Comment by Pierstoval** [comment]
:+1: on it, I might use this edition instead of Silex when I need small apps, seems good to me.

**Comment by gnugat** [comment]
Some use cases for this distribution:

* create a tailor made application (add bundles/libraries we need as opposed to the standard where we remove the ones we don't need)
* create a small application (alternative to micro frameworks)
* create a demo application (I use it in my blog examples)
* create an embeded application in third party bundle to enable testing (see https://github.com/symfony/symfony-docs/pull/5232)

**Comment by EmanueleMinotto** [comment]
Looks good to me :+1: but I would change the bin scripts, too many assumptions (like a system global `composer` and the `rm` command) imo

**Comment by mnapoli** [comment]
:+1: It's very useful. Most of the time I start a new project with Symfony I have to spend time cleaning it up from the unwanted bundles (doctrine, demo, etc.). Recent example: migrating a legacy app, I don't need Doctrine, I just want the simplest Symfony app. Additionally an "empty" distribution is very useful for learning with a simple Symfony app. The classic one is so intimidating the first time one sees it :/ (it makes Symfony look so bloated and complex when it is not).

**Comment by mickaelandrieu** [comment]
@OskarStark ok, but why ? 

Of course, I agree for an empty edition: everybody already knows what I think about the standard's one ;)

The biggest advantage of this edition is that you don't depend on the ``symfony/symfony`` metapackage and you have the choice to add the components/bundles you realy need.

Nice contribution @gnugat !

**Comment by Mandragora2** [comment]
+1 totally agree with this edition ;)

**Comment by gnugat** [comment]
Thanks everyone for your feedback!
@EmanueleMinotto: you're right, those are totally opinionated and have nothing to do here. I'll remove them.

**Comment by javiereguiluz** [comment]
I'm not against adding this empty edition, but I'd like to ask you one question and I'd like to have honest answers:

  1. You install this empty edition and start developing your application.
  2. You start installing things that the Empty Edition doesn't provide but you need for your application: Twig, Monolog, the Console component, etc.
  3. **My question is**: after having developed the real application, how different is your "empty edition + installed dependencies" from the original Symfony Standard Edition.

**Comment by mnapoli** [comment]
@javiereguiluz at step 2 you assume we all want Twig, Monolog, the Console, Doctrine, etc… The point of this edition is exactly for when you don't. So the question is moot.

> after having developed the real application

That's also an assumption that there's a fixed goal that's reached at some point. For example when migrating a legacy app written in ZF1, using ZF's data table classes and its logger, I want the smallest Symfony possible to migrate everything slowly into SF's DI container, directory layout, etc. I don't necessarily want to end up with Doctrine. And maybe rewriting every template from Smarty to Twig isn't possible right now, so why would I want Twig. There is no such thing of being "finished" in that case. Maybe in a few years it will indeed be using Twig, but who knows…

**Comment by javiereguiluz** [comment]
@mnapoli thanks for your comment. However, I don't think your use case apply for this empty edition. If you are just slowly migrating a legacy app, it may be better to pick Symfony components one by one until you reach a point when using the entire Symfony framework makes sense. This presentation by @hhamon explains it well: https://speakerdeck.com/hhamon/bringing-symfony-components-into-your-legacy-code

**Comment by mnapoli** [comment]
@javiereguiluz there are several approaches to such a migration, there's another one to what you suggest: instead of trying to change the whole of your application to *become* a Symfony app, you can introduce a Symfony app in parallel to the existing one and slowly refactor from the legacy app to the new one as well as building bridges between the 2. You can keep the 2 working at the same time thanks to stuff like [Stack](http://stackphp.com/), the HttpKernelInterface, container-interop, etc…

So in that use case, having the simplest Symfony distribution is helpful because one would want to reuse the templating engine, DB engine, etc etc from the legacy app in the new Symfony app.

**Comment by javiereguiluz** [comment]
@mnapoli fair enough ... for migrating legacy apps this empty edition may be useful. Now I want to hear from people developing new applications with this edition. Thanks.

**Comment by Pierstoval** [comment]
Actually, if I have an empty edition, I'm surely gonna build every "snippet-app" based on this edition.
Before, I used Silex, but actually it's not good enough to me to start a very small project and make it become larger in time. [Orbitale/EasyImpress](https://github.com/Orbitale/EasyImpress) is made with Silex, and I made it with it because I did not want to use Symfony because it's too big and I was lazy at that time, didn't want to take time to remove everything from the app that I don't need ; whereas having Symfony + Twig would've been all I could need.
 
Out of the context: Based on this empty edition, it could be good to create a "Symfony-app maker": You select which of the Symfony components you need, and if it includes `framework-bundle`, we just get the "Symfony Nude Edition" and update the `composer.json` to use only the specified packages. It would make a "Symfony very-light edition".
 
To summarize:
* Standard edition for classic and big projects
* Empty/Nude edition for small projects / snippets / testing

I'm totally :+1: for this

**Comment by gnugat** [comment]
Hi @javiereguiluz, thanks for your feedback. I understand your concern and I really appreciate it: if you plan to create a project with "standard" templating, mailing and database needs then you should definitely use the standard edition. The name of this distribution is very well suited.

But some projects don't need a database, some projects don't need a mailing system and some projects don't need templates. Some projects need those but might want to pick different technologies (e.g. Blade, Pomm, etc).
One of the reason that drove me to create this edition was the repeated full grown projects were we needed to strip down the standard's dependencies and then to add our own. This might not be everyone's need, but that's definitely a need.

There's also the whole micro framework niche that justifies this distribution: your arguments are all valid and they could be used against Silex, Lumen, Expressive, etc. But the truth is that some projects don't need much more than a routing system. It's not all of them, but it's sufficient to see a growing number of micro frameworks spreading everywhere. I've always thought that Symfony itself could be a micro framework, and the empty edition is the proof of it.

> **Note**: Silex is a nice microframework. But for the use cases I've encountered, I've found it not flexible enough, a bit like @Pierstoval. Since it uses Pimple, you can't reuse symfony bundles (you need to create providers), and its routing doesn't allow caching like in Symfony (one of the reason of the creation of [Flint](http://henrikbjorn.github.io/Flint/)).

Of course there's also the migration process where this distribution can be used, as mentioned by @mnapoli. There are many strategies out there, the choice really depends on the legacy application, the deadlines you have, the knowledge of the team, and how you'd like the new application to look like.

It can also be used in some small use cases: for example when creating a third party bundle, a nice way to be able to test it is to create a small embeded application (see: https://github.com/symfony/symfony-docs/pull/5232), and it's also a nice teaching material (I use it in my articles).

The Empty Edition is not a replacement for the Standard Edition. It's rather an alternative in some use cases that I've encountered many times in my experience (full grown applications, micro framework, migration and teaching material).

@Pierstoval: such a tool would definitely be great!

**Comment by gnugat** [comment]
@javiereguiluz: any news on this?

**Comment by mickaelandrieu** [comment]
> it could be good to create a "Symfony-app maker"
be patient ;)

**Comment by gnugat** [comment]
Closed, due to lack of feedback
